dot/sync/unready_blocks.go

```diff
@@ -12,16 +13,80 @@ import (
 	"github.com/ChainSafe/gossamer/lib/common"
 )
 
+type Fragment struct {
+	chain []*types.BlockData
+}
+
+func NewFragment(chain []*types.BlockData) *Fragment {
+	return &Fragment{chain}
+}
+
+func (f *Fragment) Filter(p func(*types.BlockData) bool) *Fragment {
+	filtered := make([]*types.BlockData, 0, len(f.chain))
+	for _, bd := range f.chain {
+		if p(bd) {
+			filtered = append(filtered, bd)
+		}
+	}
+	return NewFragment(filtered)
+}
+
+// Find return the first occurrence of a types.BlockData that
+// satisfies the predicate p
+func (f *Fragment) Find(p func(*types.BlockData) bool) *types.BlockData {
+	for _, bd := range f.chain {
+		if p(bd) {
+			return bd
+		}
+	}
+
+	return nil
+}
+
+func (f *Fragment) Last() *types.BlockData {
+	if len(f.chain) > 0 {
+		return f.chain[len(f.chain)-1]
+	}
+
+	return nil
+}
+
+func (f *Fragment) Len() int {
+	return len(f.chain)
+}
+
+func (f *Fragment) Iter() iter.Seq[*types.BlockData] {
+	return func(yield func(*types.BlockData) bool) {
+		for _, bd := range f.chain {
+			yield(bd)
```

Comment:
We should check the yield function result

```suggestion
			if !yield(bd){
			    return
			}
```